src/thread.rs

```diff
@@ -0,0 +1,66 @@
+use std::fmt::{self, Debug};
+
+#[cfg(syn_can_use_thread_id)]
+mod imp {
+    use std::thread::{self, ThreadId};
+
+    /// ThreadBound is a Sync-maker and Send-maker that allows accessing a value
+    /// of type T only from the original thread on which the ThreadBound was
+    /// constructed.
+    pub struct ThreadBound<T> {
+        value: T,
+        thread_id: ThreadId,
+    }
+
+    unsafe impl<T> Sync for ThreadBound<T> {}
+
+    // Send bound requires Copy, as otherwise Drop could run in the wrong place.
+    unsafe impl<T: Copy> Send for ThreadBound<T> {}
+
+    impl<T> ThreadBound<T> {
+        pub fn new(value: T) -> Self {
+            ThreadBound {
+                value: value,
+                thread_id: thread::current().id(),
+            }
+        }
+
+        pub fn get(&self) -> Option<&T> {
+            if thread::current().id() == self.thread_id {
+                Some(&self.value)
+            } else {
+                None
+            }
+        }
+    }
+}
+
+#[cfg(not(syn_can_use_thread_id))]
```

Comment:
I think we can do better than this if we don't have `std::thread::Thread::id`, just with some performance overhead. This is safe, as we don't implement `Send` and `Sync` on the type, but perhaps unfortunate.

We can "emulate" ThreadId's behaviour using atomics and thread locals. If you decide not to bother with that, please add a comment explaining what restrictions are put on people using old rustc versions ;-).

```rust
#[cfg(syn_can_use_thread_id)]
mod thread_id {
    use std::thread;

    pub use std::thread::ThreadId;
    pub fn current() -> ThreadId {
        thread::current().id()
    }
}

#[cfg(not(syn_can_use_thread_id))]
mod thread_id {
    use std::sync::atomic::{AtomicUsize, Ordering, ATOMIC_USIZE_INIT};

    static NEXT_THREAD_ID: AtomicUsize = ATOMIC_USIZE_INIT;
    thread_local! {
        static THREAD_ID: usize = NEXT_THREAD_ID.fetch_add(1, Ordering::Relaxed);
    }

    pub type ThreadId = usize;
    pub fn current() -> ThreadId {
        THREAD_ID.with(|id| id)
    }
}
```

Reply:
Nice! Done.